### ISSUE TITLE
Constraint missing (async function must return one of these types) in draft C# specification

### ISSUE THREAD

**Issue by gafter** [dotnet/roslyn#14001]
The following program
```cs
using System;
class C
{
    static void Main()
    {
        // error CS4010: Cannot convert async anonymous method to delegate type 'System.Func'. An async anonymous method may return void, Task or Task, none of which are convertible to 'System.Func'.
        Func<int> x = async delegate {
            throw null;
        };
    }
}
```
Is rejected by the compiler.

Seeking guidance from the language spec, I could not find anything to forbid this program. Perhaps there is a constraint missing in 6.5 (“If F is async, D has the return type void, Task, or Task for some T.”) or misworded?

**Comment by acelent** [comment]
How should this not be forbidden?  None of the implicit `async` return types (`void`, `Task` or `Task<TResult>`) is compatible with `int`.

**Comment by axel-habermaier** [comment]
@acelent: Yes, that's why @gafter thinks the specification must be fixed. In order words: The compiler is right, the spec is wrong.

**Comment by acelent** [comment]
@axel-habermaier, I understand that.

I'm looking at the specification.  The ***inferred return type*** is not defined until §7.5, and anonymous function expressions in §7.15 which points to §10.14, it should be §10.15, which states that `void`, `Task` and `Task<T>` are the valid async method return types.

However, §6.5 already points to §7.15, and a couple of §7.15's subchapters point back to §6.5.

Perhaps what should be discussed is not the lack of constraints, but if repeating/reminding/introducing the valid async return types is useful in §6.5, more than just linking to other chapters.
